### DISCUSSION TITLE
Options for passing data from pre-render script to individual qmds in a quarto website

### DISCUSSION THREAD

**Discussion by boshek** [quarto-dev/quarto-cli#7300] [Q&A] [answered]
### Description

How are folks passing data generated from an expensive step that should only happen once but is used my multiple qmds in your website? Example: 

So if in my `_quarto.yml` file I specify a `pre-render` script called `process.R` that looks like this:

```r
data <- make_expensive_api_call()
write_rds(data, "data.rds")
```

And then in various `qmd` in my webpage I have a repeated line like this:

````
```r
data <- readRDS("data.rds")
```
````

And then in my post-render I have a clean script that deletes `data.rds` file. 

Ignoring my choice of `rds`, are there other approaches to this? I mean this is probably fine but I just wanted to see if folks had different solutions to this.

**Comment by mcanouil** [comment]
I don't think there is another approach or use `targets` (which basically does that but better).

**Reply by boshek** [reply]
Sorry to let this languish. I am interested in figuring out a better polyglot solution for this. So a question - can you point me to where I might understand the environment for the pre-render scripts? What is the lifecycle of the memory for that process? Is it destroyed once the rendering is started? Thanks!

**Reply by mcanouil** [reply]
This is in the documentation.
It's simply a script launcher either before and/or after rendering.
Everything in those scripts is entirely up to you.

https://quarto.org/docs/projects/scripts.html#pre-and-post-render

**Reply by boshek** [reply]
Thanks I did review those. I don't see anything about the environment in there, apologies if I am missing something obvious. Like is that the same session that actually does the rendering or a separate session? When is it destroyed etc?

**Reply by mcanouil** [reply]
There is currently no Quarto internal guide.
I am not sure but the "session" (if you mean bash/powershell) is the same since you have access to some Quarto environment variables you can use in pre/post- render scripts.
Since those are not available outside, anything ran as part of Quarto rendering is likely to be "destroyed" when exiting rendering.

Maybe @cscheid can help here.

**Reply by cscheid** [reply, marked as answer]
> Ignoring my choice of rds, are there other approaches to this? I mean this is probably fine but I just wanted to see if folks had different solutions to this.

There's no persistence from the execution of one .qmd to the next, aside from artifacts your code might create as a sideeffect (in the filesystem, databases, web services, git repos, etc). So you will need, somewhere in each of your .qmd files, to actually load them. If you're concerned about code repetition, you can consider using an include shortcode that contains the executable code you want to run, so that the repetition is condensed into a single line.
